src/main/java/org/mockito/internal/junit/JUnitRule.java

```diff
@@ -14,25 +13,74 @@
 public class JUnitRule implements MockitoRule {
 	
 	private final MockitoLogger logger;
+    private final boolean silent;
 
-	public JUnitRule(MockitoLogger logger) {
+    public JUnitRule(MockitoLogger logger, boolean silent) {
 		this.logger = logger;
-	}
+        this.silent = silent;
+    }
+
 	@Override
 	public Statement apply(final Statement base, FrameworkMethod method, final Object target) {
-		return new Statement() {
-            @Override
-            public void evaluate() throws Throwable {
-                WarningsCollector c = new WarningsCollector();
-                MockitoAnnotations.initMocks(target);
-                try {
-                    base.evaluate();
-                } catch(Throwable t) {
-                    logger.log(c.getWarnings());
-                    throw t;
-                }
-                Mockito.validateMockitoUsage();
+        if (silent) {
+            return new SilentStatement(target, base);
+        } else {
+            String testName = target.getClass().getSimpleName() + "." + method.getName();
+            return new DefaultStatement(target, testName, base);
+        }
+    }
+
+    public JUnitRule silent() {
+        return new JUnitRule(logger, true);
+    }
+
+    private class SilentStatement extends Statement {
+        private final Object target;
+        private final Statement base;
+
+        public SilentStatement(Object target, Statement base) {
+            this.target = target;
+            this.base = base;
+        }
+
+        public void evaluate() throws Throwable {
+            MockitoAnnotations.initMocks(target);
+            base.evaluate();
+            Mockito.validateMockitoUsage();
+        }
+    }
+
+    private class DefaultStatement extends Statement {
+        private final Object target;
+        private final String testName;
+        private final Statement base;
+
+        DefaultStatement(Object target, String testName, Statement base) {
+            this.target = target;
+            this.testName = testName;
+            this.base = base;
+        }
+
+        public void evaluate() throws Throwable {
+            RuleStubbingHintsReporter reporter = new RuleStubbingHintsReporter();
+            Mockito.framework().addListener(reporter);
+            try {
+                performEvaluation(reporter);
+            } finally {
+                Mockito.framework().removeListener(reporter);
+            }
+        }
+
+        private void performEvaluation(RuleStubbingHintsReporter reporter) throws Throwable {
+            MockitoAnnotations.initMocks(target);
+            try {
+                base.evaluate();
+            } catch(Throwable t) {
+                reporter.getStubbingArgMismatches().format(testName, logger);
+                throw t;
             }
-        };
+            reporter.getUnusedStubbings().format(testName, logger);
```

Comment:
Would it be possible to let the reporter print everything i.e. stubbing arg mismatches and/or unused stubbings, instead of making two calls ?
This would allow to keep the stubbing reporting concerns in the reporting concerns instead of having them in the statement evaluator.